File: examples/ch13/exercise/shipping.py
"""Chapter 10 exercise starter: the shipping-cost tangle.

Refactor with your agent using the "Strategy, right-sized" prompt from the
chapter (adapt the names), then grade the output with the review checklist.
"""

from dataclasses import dataclass


@dataclass(frozen=True)
class Shipment:
    weight_kg: float
    destination_country: str
    declared_value: float


def shipping_cost(shipment: Shipment, carrier: str) -> float:
    """Return the shipping cost for the given carrier option."""
    if carrier == "standard":
        return 4.90 + 0.50 * shipment.weight_kg
    elif carrier == "express":
        base = 9.90 + 1.20 * shipment.weight_kg
        if shipment.declared_value > 100.0:
            base += 2.50  # mandatory insurance
        return base
    elif carrier == "pickup":
        return 0.0
    elif carrier == "international":
        if shipment.destination_country in ("DE", "FR", "NL"):
            return 14.90 + 1.00 * shipment.weight_kg
        return 24.90 + 2.20 * shipment.weight_kg
    else:
        raise ValueError(f"unknown carrier: {carrier}")

File: examples/ch13/ts/exercise/shipping.ts
// Chapter 13 exercise starter: the shipping-cost tangle.
//
// Refactor with your agent using the "Strategy, right-sized" prompt from the
// chapter (adapt the names), then grade the output with the review checklist.

export interface Shipment {
  readonly weightKg: number;
  readonly destinationCountry: string;
  readonly declaredValue: number;
}

const EU = new Set(["DE", "FR", "NL"]);

export function shippingCost(shipment: Shipment, carrier: string): number {
  switch (carrier) {
    case "standard":
      return 4.9 + 0.5 * shipment.weightKg;
    case "express": {
      let base = 9.9 + 1.2 * shipment.weightKg;
      if (shipment.declaredValue > 100) base += 2.5; // mandatory insurance
      return base;
    }
    case "pickup":
      return 0;
    case "international":
      return EU.has(shipment.destinationCountry)
        ? 14.9 + 1.0 * shipment.weightKg
        : 24.9 + 2.2 * shipment.weightKg;
    default:
      throw new Error(`unknown carrier: ${carrier}`);
  }
}
